inference-engine/include/details/ie_mmap_allocator.hpp

```diff
@@ -0,0 +1,276 @@
+// Copyright (C) 2018-2020 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+/**
+ * @brief The header file defines utility PreAllocator class
+ * 
+ * @file ie_pre_allocator.hpp
+ */
+#pragma once
+
+#include <details/ie_exception.hpp>
+#include <memory>
+
+#include "ie_allocator.hpp"
+
+#ifdef _WIN32
+#include <windows.h>
+#define MAP_FAILED nullptr
+#else
+#include <unistd.h>
+#include <fcntl.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#define INVALID_HANDLE_VALUE -1
+#endif
+
+namespace InferenceEngine {
+namespace details {
+/*
+ * @brief This is a helper class to wrap memory mapped files
+ */
+class MmapAllocator : public IAllocator {
+public:
+    using Ptr = std::shared_ptr<MmapAllocator>;
+
+#if defined(ENABLE_UNICODE_PATH_SUPPORT) && defined(_WIN32)
+    using path_type = std::wstring;
+#else
+    using path_type = std::string;
+#endif
+
+    MmapAllocator(const path_type& path, size_t offset = 0, LockOp lock = LOCK_FOR_READ)
+        : _path(path), _offset(offset), _size(0), _lock(lock), _data(nullptr) {
+    }
+
+    /**
+     * @brief Locks a handle to heap memory accessible by any memory manipulation routines
+     * @return The generic pointer to a memory buffer
+     */
+    void* lock(void* handle, LockOp) noexcept override {
+        return handle;
+    }
+    /**
+     * @brief The MmapAllocator class does not utilize this function
+     */
+    void unlock(void*) noexcept override {}  // NOLINT
+
+    /**
+     * @brief Returns a pointer to mapped memory
+     * @param size Size in bytes
+     * @return A handle to the preallocated memory or nullptr
+     */
+    void* alloc(size_t size) noexcept override {
+        map(_path.c_str(), _offset, size, _lock);
+        return _data;
+    }
+    /**
+     * @brief The PreAllocator class cannot release the handle
+     * @return false
+     */
+    bool free(void*) noexcept override {  // NOLINT
+        unmap();
+        closeFile();
+        return true;
+    }
+
+    /**
+     * @brief Deletes current allocator.
+     * Can be used if a shared_from_irelease pointer is used
+     */
+    void Release() noexcept override {
+        free(nullptr);
+        delete this;
+    }
+
+    size_t size() {
+        return _size;
+    }
+
+protected:
+    virtual ~MmapAllocator() = default;
```

Comment:
We can avoid or detect possible memory leaks if implement destructor.

Reply:
I would suggest to extract `closeFile()`, `map()`, `unmap()` and related data to a separate class/interface (e.g. `Mmapper`) and implement Linux and Windows versions in separate `.cpp` files.  
Additionally, implementing RAII pattern in `Mmapper`  and proper ownership, would simplify prevention of resources/memory leakage. For example:
```cpp
class Mmapper {
publlic:
    Mmapper(const path_type& path, size_t offset, size_t size, LockOp lock) {
         // TODO all memory mapping stuff
    }

    ~Mmapper() {
         unmap();
         CloseFile();
    }
    void* getData() const noexcept { return _data; }
...
};
...
class MmapAllocator {
...
void* alloc(size_t size) noexcept override {
  try {
    _mapper.reset(new Mmapper(_path.c_str(), _offset, size, _lock));
    return _mapper->getData();
  } catch(const std::exception& ex) {
    // TODO log the error
    _mapper.reset();
    _return nullptr;
  }
}
bool free() noexcept override {
    _mapper.reset();
    return true;
}
...
private:
    std::unique_ptr<Mmapper> _mapper;
}
```